utils/s3/client.cc

```
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+        auto in = std::move(in_);
+        auto body = co_await util::read_entire_stream_contiguous(in);
+        _upload_id = parse_multipart_upload_id(body);
+        if (_upload_id.empty()) {
+            co_await coroutine::return_exception(std::runtime_error("cannot initiate upload"));
+        }
+        s3l.trace("created uploads for {} -> id = {}", _object_name, _upload_id);
+    });
+}
+
+future<> client::upload_sink::upload_part(unsigned part_number, memory_data_sink_buffers bufs) {
+    s3l.trace("PUT part {} {} bytes in {} buffers (upload id {})", part_number, bufs.size(), bufs.buffers().size(), _upload_id);
+    auto req = http::request::make("PUT", _client->_host, _object_name);
+    req._headers["Content-Length"] = format("{}", bufs.size());
+    req.query_parameters["partNumber"] = format("{}", part_number);
+    req.query_parameters["uploadId"] = _upload_id;
+    req.write_body("bin", bufs.size(), [this, part_number, bufs = std::move(bufs)] (output_stream<char>&& out_) mutable -> future<> {
+        auto out = std::move(out_);
+        s3l.trace("upload {} part data (upload id {})", part_number, _upload_id);
+        for (auto&& buf : bufs.buffers()) {
+            co_await out.write(buf.get(), buf.size());
+        }
+        co_await out.flush();
+        co_await out.close();
+    });
+
+    auto units = co_await get_units(_flush_sem, 1);
+    (void)_http.make_request(std::move(req), [this, part_number] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
+        auto etag = rep.get_header("ETag");
+        s3l.trace("uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+        _part_etags[part_number] = std::move(etag);
+        return make_ready_future<>();
+    }).handle_exception([this, part_number] (auto ex) {
+        s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
```

(and need to fix error handling)

---

Can you elaborate? If the part uploading fails there appears a "hole" in the `_part_etags`, and at the end there can be two outcomes:
* flush() is called, it notices the hole and aborts the upload
* close() is called w/o flush, it notices the upload is not flushed and aborts the upload as well

---

If it works, that's fine, but a swallowed exception needs a comment explaining why it's okay.

It would also be good to test failures, though maybe hard. Maybe setup minio with a quota and upload something too large?

---

S3 protocol is very fragile and it's easy to trigger server error by (unwillingly) violating any of 132645 corner cases. I've a patch that adds random delays to upload-parts that I used to test upload-finish synchronization. I'll think towards adding "error injection" too